docs/proposals/webhook-validation.md

```diff
@@ -0,0 +1,130 @@
+
+---
+title: webhook-validation
+authors:
+  - "@ImJasonH"
+reviewers:
+  - "@gmontero"
+  - "@zhangtbj"
+approvers:
+  - "@qu1queee"
+  - "@adamkaplan"
+creation-date: 2020-03-19
+last-updated: 2020-03-19
+status: provisional
+---
+
+# Webhook Validation
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Open Questions [optional]
+
+Do we want to require that operators run another job to synchronously validate client requests?
+
+Do we want to rely on Knative's webhook packages, or write our own code to manage certs ourselves?
+Is there anything already out there in controller-gen we should be using?
+
+## Summary
+
+Kubernetes provides facilities to inject synchronous validation of CRUD operations from clients for specified types, such as Shipwright's CRD types.
+
+Webhooks can also mutate incoming requests, including setting default values and converting between versions of CRDs. These features should be considered later, and are not described in this proposal.
+
+We should take advantage of these features and include a webhook validation deployment in the Shipwright installation to provide this functionality.
+
+## Motivation
+
+Today, when a user creates a resource (e.g., a Build, BuildRun, etc.), they are immediately stored in the cluster's etcd storage, and the Shipwright controller begins reconciling the object.
+
+If the controller finds that the resource is invalid, it updates the resource's `.status` to indicate this state and stops reconciling it.
+
+Examples of invalid resources include:
+
+- a BuildRun referencing a Build that doesn't exist
+- a BuildRun referencing a ServiceAccount that doesn't exist
+- a Build referencing a (Cluster)BuildStrategy that doesn't exist
+- a BuildStrategy containing zero steps
+
+(This is a partial list, and new validation can be assumed to be added over time as the Shipwright API evolves.)
+
+By allowing invalid resources to be accepted and stored, we require clients to watch or poll all resources to ensure their request was valid, and we add to the API surface of resources that otherwise would not need to report a status, like a Build.
+
+Always accepting and storing resources, even invalid ones, can also lead to additional storage of known-invalid objects, which clutter the user experience and can lead to confusion, and in extreme cases, cluster resource exhaustion.
+
+Mixing reconciliation and validation code in the controller also complicates our codebase, leading to PRs like https://github.com/shipwright-io/build/pull/641 which add new validation-centric statuses that the user must watch for.
+As we add new validation scenarios, we currently add new `Reason` strings that clients must be aware of, which is cumbersome and won't scale indefinitely.
```

Comment:
Imo this would not change with a webhook. We will have different validations and should continue to report the problem as specific as possible, for example using an error code that would be very much the same as the reason that we have today. A good client would have to provide the user different assistance depending on what the error code is.

Reply:
A goal (perhaps unstated) of this proposal is to not have to define and maintain a table of all possible validation error scenarios and associated unique codes/reasons. That doesn't scale well much beyond where we're already at, and unless we invest in generating documentation, we _will_ have them fall out of sync eventually. Incorrect or incomplete documentation can be worse than no documentation.

Instead, I'd like to invest in human-readable descriptive error messages that we can proactively put in front of users' faces. Those errors can _also_ include codes (i.e., along the lines of HTTP 400 vs 404) to classify the kinds of errors, if we want.

Reply:
I think it will change @SaschaSchwarze0 , that does not mean we will get rid of how we populate our Status.Conditions, we still need the core states of a resource that runs to completion, e.g. `True`, `False`, `Unknown` and all potential flavors of the `Reason` field. What we will do is reduce some of those flavors, e.g. `BuildStrategyNotFound` and allow the admission webhook to block the creation of a BuildRun if the strategy is not there.

@ImJasonH I think the table of error scenarios is still needed till some extent, probably an smaller table of error scenarios but we still need to decide in the logic around when to mark something as `False` with a `Failed` Reason, or with a `BuildRunTimeout` one. I dont see this going away.

Also, I´m not sure if the error categorization discussion is out of the scope on this PR, maybe too much of an implementation detail, which we might want to take somewhere else. I do have some strong opinions around entering the business of parsing errors.

Reply:
Good discussion. I was mainly focusing on those validations that we would perform before a resource gets created (like referenced secrets and buildstrategy when a Build creation happens). I agree that we will have HTTP status codes which can help us to categorize errors already. Though, I guess we will have many errors that translate into the same HTTP status codes. Not just because of that, I think that errors should be categorized using a computer-readable error code (like the current reasons we have for Builds). I am definitely not against human-readable descriptions. But, those have a tendency to be changed from time to time, and client interfaces - especially localized one's - typically cannot rely on the human-readable description. They need other means to identify an error to guide their users to make the necessary correction.